libbeat/common/mapval/path.go

```diff
@@ -32,11 +32,21 @@ type PathComponentType int
 
 const (
 	// PCMapKey is the Type for map keys.
-	PCMapKey = iota
+	PCMapKey PathComponentType = iota
 	// PCSliceIdx is the Type for slice indices.
 	PCSliceIdx
 )
 
+func (pct PathComponentType) String() string {
+	if pct == PCMapKey {
+		return "map"
+	} else if pct == PCSliceIdx {
+		return "slice"
+	} else {
+		panic(fmt.Sprintf("Unknown path component type! %d", int(pct)))
```

Comment:
Let's use an error here instead of a panic. What if someone uses this for production code and is not aware this can apnice?

Reply:
Well, this practically cannot panic. The only way that would happen is if someone was messing deep in the guts of mapval and added a new type without overriding this.

I would say then, that if this did panic, it'd be a bug in mapval itself, not in user-facing code. Checking invariants like that seems like the right place for panic no?

Reply:
I'm realizing this is similar to the other panic described earlier. Any case where another type is encountered would definitely be a programming error. You would need to create a new `PathComponentType` with a different (manual) value. 

@urso I see you +1'd this change, but opposed removing the other panic. Maybe you can help me understand the difference?

Reply:
Assuming mapval package could be used in other places in beats, the mapval package might be used to validate any kind of external input. We might even consider users to configure validation (based on mapval) in the heartbeat config file. When working with external input, mapval should not panic.

If the panic can only indicate programming errors, I'm fine with the panic. If the panic can be triggered by external (yet unchecked) input then we should go with an error. mapval has to check input, so we must assume any kind of invalid input.

It's common in String methods to not create a panic, but return an 'Unknown'-string.

Btw. consider using stringer like:

```

//go:generate stringer -type=PathComponentType -linecomment=true

const (
	PCMapKey PathComponentType = 1 + iota // map

	// PCSliceIdx is the Type for slice indices.
	PCSliceIdx                            // slice
)
```

Notice how I set `PCMapKey` to 1. That is, the zero-value is invalid. Oftentimes one defines an 'invalid'-const to be 0 like:

```

//go:generate stringer -type=PathComponentType -linecomment=true

const (
    PCMapInvalid PathComponentType = iota // invalid
	PCMapKey                              // map
	PCSliceIdx                            // slice
)
```

Reply:
Thanks for the feedback, I'll take your advice and return an unknown string. I think that's a great compromise.

Also, thanks for the tip on the 1+, makes a lot of sense.